transport: replace throwing protocol_exception with returns

DarioMirovic opened the pull request:
Replace throwing `protocol_exception` with returning it as a result or an exceptional future in the transport server module. The goal is to improve performance.

Most of the `protocol_exception` throws were made from `fragmented_temporary_buffer` module, by passing `exception_thrower()` to its `read*` methods. `fragmented_temporary_buffer` is changed so that it now accepts an exception creator, not exception thrower. `fragmented_temporary_buffer_concepts::ExceptionCreator` concept replaced `fragmented_temporary_buffer_concepts::ExceptionThrower` and all methods that have been throwing now return failed result of type `utils::result_with_eptr`. This change is then propagated to the callers.

The scope of this patch is `protocol_exception`, so commitlog just calls `.value()` method on the result. If the result failed, that will throw the exception from the result, as defined by `utils::result_with_eptr_throw_policy`. This means that the behavior of commitlog module stays the same.

transport server module handles results gracefully. All the caller functions that return non-future value `T` now return `utils::result_with_eptr<T>`. When the caller is a function that returns a future, and it receives failed result, `make_exception_future(std::move(failed_result).value())` is returned. The rest of the callstack up to the transport server `handle_error` function is already working without throwing, and that's how zero throws is achieved.

cql3 module changes do the same as transport server module.

Fixes: #24567

This PR is a continuation of #24738 [transport: remove throwing protocol_exception on connection start](https://github.com/scylladb/scylladb/pull/24738). This one fixes the issue which the previous PR just referenced. It should be backported to the same versions: 2025.3, 2025.2, 2025.1.

DarioMirovic commented:
@yaronkaikov `call-jira-status-in-progress` failed with errors like:
```
/home/runner/work/_temp/f94fd10e-4e39-4ee8-9c84-56da092c5976.sh: line 14: protocol_exception: command not found
```

mykaul commented:
> @yaronkaikov `call-jira-status-in-progress` failed with errors like:
> 
> ```
> /home/runner/work/_temp/f94fd10e-4e39-4ee8-9c84-56da092c5976.sh: line 14: protocol_exception: command not found
> ```

@dani-tweig ?

dani-tweig commented:
> > @yaronkaikov `call-jira-status-in-progress` failed with errors like:
> > ```
> > /home/runner/work/_temp/f94fd10e-4e39-4ee8-9c84-56da092c5976.sh: line 14: protocol_exception: command not found
> > ```
> 
> @dani-tweig ?

@mykaul , I have added the call action of changing the status in Jira to scylladb, since there are already PRs that refer to projects that exist only in Jira (VECTOR). 
While running the actions in scylladb, I found a bug that had not been revealed until now, and I fixed it.

avikivity commented on utils/result.hh:
```diff
@@ -59,4 +59,24 @@ concept ResultRebindableTo =
 template<typename T, ExceptionContainerResult R>
 using rebind_result = bo::result<T, typename R::error_type, exception_container_throw_policy>;
```
> Refs: 24567

bad syntax

ptrsmrn commented on utils/result.hh:
```diff
@@ -59,4 +59,24 @@ concept ResultRebindableTo =
+        }
+    }
+};
+
```
@piodul will be back only in 1 week.

ptrsmrn replied:
We will.

ptrsmrn reviewed (approved):
One concern I have is that the code didn't get prettier because of these changes, but I don't see a way how to hide all this recurring boilerplate (unless via macro).

DarioMirovic commented:
> One concern I have is that the code didn't get prettier because of these changes, but I don't see a way how to hide all this recurring boilerplate (unless via macro).

True. Either macros with standardized return if error, or more verbose code it is. And this is a solution that uses `std::exception_ptr`, not requiring specifying exception types that can happen through the callstack.

piodul commented on transport/server.cc:
```diff
@@ -1694,21 +1751,34 @@ cql_server::connection::make_schema_change_event(const event::schema_change& eve
-        auto message = response->make_message(_version, compression);
-        message.on_delete([response = std::move(response)] { });
-        return _write_buf.write(std::move(message)).then([this] {
+        utils::result_with_eptr<scattered_message<char>> message = response->make_message(_version, compression);
+        if (!message) [[unlikely]] {
+            if (auto* exp = try_catch<exceptions::cassandra_exception>(message.error())) {
+                clogger.error("{}: write_response: failed to make message, code {}, message [{}]", _client_state.get_remote_address(), exp->code(), exp->what());
+            } else if (auto* exp = try_catch<std::exception>(message.error())) {
+                clogger.error("{}: write_response: failed to make message, message [{}]", _client_state.get_remote_address(), exp->what());
+            } else {
+                clogger.error("{}: write_response: failed to make message", _client_state.get_remote_address());
+            }
+            return make_ready_future<>();
+        }
+        message.value().on_delete([response = std::move(response)] { });
+        return _write_buf.write(std::move(message).value()).then([this] {
```
You have snuck in a logic change into a refactoring commit. Please move it out to a separate commit.

By the way, why the change? Why do we swallow the exception now rather than propagating it outwards?

DarioMirovic replied:
I'll extract `make_message` and `make_frame` changes into a separate commit. As for the logic change, I see, I should keep it the same. There seem to be 13 places in `transport/event_notifier.cc` that call `write_response`. I'll propagate the error further up without rethrowing for both transport server and transport event notifier.

piodul commented on utils/result.hh:
```diff
@@ -59,4 +59,24 @@ concept ResultRebindableTo =
 using rebind_result = bo::result<T, typename R::error_type, exception_container_throw_policy>;
 
+struct result_with_eptr_throw_policy : bo::policy::base {
+    template<class Impl> static constexpr void wide_value_check(Impl&& self) {
+        if (!base::_has_value(self)) {
+            std::rethrow_exception(base::_error(self));
+        }
+    }
```
The fact that boost::outcome can be configured to throw when `value()` is called is nice, but unfortunately std::expected does not work like that and only throws `std::bad_expected_access<ErrorType>` :/

This is a preexisting issue, no need to do anything here but worth keeping in mind if we eventually decide to move to the standard library-native solution.

avikivity commented on utils/result.hh:
```diff
@@ -59,4 +59,24 @@ concept ResultRebindableTo =
 template<typename T, ExceptionContainerResult R>
 using rebind_result = bo::result<T, typename R::error_type, exception_container_throw_policy>;
 
+struct result_with_eptr_throw_policy : bo::policy::base {
```
Objecting to the acronym-heavy name.

avikivity commented on transport/request.hh:
```diff
@@ -71,130 +73,237 @@ public:
+        if (!v) [[unlikely]] {
+            return bo::failure(std::move(v).error());
+        }
+        return wire_to_consistency(v.value());
```
I think the patch goes a little bit too far. There are expected exceptions like bad protocol numbers (since they are part of the negotiation). There are unexpected exceptions (like failing to read a short or something) since drivers usually behave.

There will be some extra instructions involved in all this checking, so some performance impact.

avikivity commented on transport/server.cc:
```diff
@@ -992,8 +997,11 @@ void cql_server::connection::update_scheduling_group() {
-    auto challenge = sasl_challenge->evaluate_response(buf);
+    utils::result_with_eptr<bytes_view> buf = in.read_raw_bytes_view(in.bytes_left());
+    if (!buf) {
+        return make_exception_future<std::unique_ptr<cql_server::response>>(std::move(buf).error());
```
Why is this path left throwing? Not that I think it's important. But I want to understand the pattern.

DarioMirovic replied:
It's returning an exception future, which is then handled without rethrows in `process_request_one` where `futurize_invoke` will now return from its `try` block instead of catching thrown exception and returning from there.

avikivity commented on transport/request.hh:
```diff
@@ -71,130 +73,237 @@ public:
+        if (!v) [[unlikely]] {
+            return bo::failure(std::move(v).error());
+        }
+        return be_to_cpu(v.value());
```
IIRC boost::outcome::result::value() is a checking version and there is a non-checking assume_value() variant too. Here probably the compiler will figure out that the previous check implies it can skip the second exception check, but please go over and see if there are places where we call .value(), we know it's not an exception, but the check is too far away. These are candidates for using assume_value().

DarioMirovic replied:
That's true, thanks for the suggestion! I'll update the code.

avikivity commented:
I'd like to test the performance of this change. Normally I'd ask for perf-simple-query, but it runs queries through query_processor directly and bypasses transport. It would be better to have go through TCP (with a fake driver to drive traffic). However that's too much extra work for this patch.

@mykaul let's plan to enhance our low-level performance tests.

DarioMirovic commented:
In V2:
- split the main commit that removes `protocol_exception` throws into several smaller ones
- changed the way `write_response` handles `make_message` errors so that the logic is not changed, just the error propagation mechanism
- replaced `.value()` with `.assume_value()` and `.error()` with `.resume_error()` after the check to avoid double check
    - kept `.value()` in tests to keep strong checks
    - kept `.value()` in commitlog.cc to keep the old behavior and not propagate changes

DarioMirovic commented:
In V3:
- Renamed `result_with_eptr` to `result_with_exception_ptr`
- Added 6 tests that test failure cases where errors happen, but not C++ exceptions

avikivity commented:
I think it can be measured.

1. Start a single-node single-shard cluster under `perf stat`.
2. Run a saturating (but not too stressful) workload. Can be as simple as reading a fully cached cassandra-stress workload. The simpler the workload, the more accurate the result, but the shard has to reach capacity. 2M ops (total) should be enough. No errors.
3. Stop the cluster
4. Record perf-stat output
5. Divide total instructions by total ops to measure instructions per op. The variations from startup would be amortized by the 2M ops.
6. Repeat pre and post patch.

Better configure with --mode=release --disable-lto --use-profile='' so we don't skew the results due to the pre-trained profile.

I expect some regression, but let's see how much. Taking numbers out of a hat, I'll accept 500 insn/op or below and we can negotiate up to 1000.

mykaul commented:
Write before vs. after system stats are interestingly very very different:
real=5:12.37  user=27.461  sys=3:54.53  cpu=83%
real=5:12.82  user=4:10.79  sys=11.530  cpu=83%

avikivity commented:
Oh and the results are even in the cover letter.

Are the results for release more? 100k insn/op look bad compared to perf-simple-query, I wouldn't have guessed so much overhead in transport/.

DarioMirovic commented:
> > Write before vs. after system stats are interestingly very very different: real=5:12.37 user=27.461 sys=3:54.53 cpu=83% real=5:12.82 user=4:10.79 sys=11.530 cpu=83%
> 
> These are quite different. I will rerun once more to verify.

I have run the benchmark again, multiple times for both read and write. All the results gave between 5 and 56 seconds of sys time, except that once the one with my patch reported 3:35. I am not sure what this anomaly is. All the other values are in line with what we have observed so far and what I reported here.

avikivity commented:
Looks good, but see my xfail comment.

DarioMirovic commented:
@avikivity I have executed `perf-simple-query` benchmarks too. Since it's a bit longer now, I moved `perf-cql-raw` and `perf-simple-query` benchmark results into two `<details>` collapsable blocks.

avikivity commented:
> @avikivity I have executed `perf-simple-query` benchmarks too. Since it's a bit longer now, I moved `perf-cql-raw` and `perf-simple-query` benchmark results into two `<details>` collapsable blocks.

bypass-cache etc don't make sense for this benchmark, since the sstable parsing will completely dominate over any overhead you add.

perf-simple-query doesn't measure transport/, it's irrelevant. It should be replaced by perf-cql-raw (we can keep perf-simple-query, but for general use, the new test is better).

DarioMirovic commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -206,6 +225,23 @@ def test_unexpected_message_during_auth(scylla_only, no_ssl, host):
     assert cpp_exception_metrics_after - cpp_exception_metrics_before <= cpp_exception_threshold, "Expected C++ protocol errors to not increase"
 
+# Test if the error is raised when sending an unknown compression algorithm.
+def test_unknown_compression_algorithm(scylla_only, no_ssl, host):
```
Oh, that. In this PR the first commit is an additional test for the changes made in the first `protocol_exception` PR. I added that test case here to avoid creating separate PR, but put it as the first commit since it is not affected by the `protocol_exception` code changes from later commits in this patch series.

avikivity replied:
ok

nuivall commented:
@DarioMirovic is 5% write throughput degradation reproducible?

DarioMirovic commented:
> @DarioMirovic is 5% write throughput degradation reproducible?

I have repeated the benchmark with clean state - reboot computer, put in performance mode, rebuild, closed other apps that might affect CPU and disk usage.

run count: 5 times before and 5 times after the patch
duration: 300 seconds

Average write throughput median before patch: 41155.99
Average write throughput median after patch: 42193.22

Median absolute deviation is also lower now, with values in range 350-550, while the previous runs' values were in range 750-1350.

mykaul commented:
@DarioMirovic - https://www.alexgallego.org/perf/compiler/explorer/flatbuffers/smf/2018/06/30/effects-cpu-turbo.html is a good read.

DarioMirovic commented:
@mykaul thank you for the article link. I set up the environment and retested.

Here are the results with those measures.
cc @avikivity @nuivall 

Environment:
CPU min frequency 2GHz
CPU max frequency 2GHz
CPU governor performance

I have checked that the CPU cores are always at 2GHz, not deviating from that value neither up nor down.

I run the tests 5 times * 300 seconds each run, before and after the patch.
Number of operations: 5 runs * 300 seconds per run * ~20000 ops/sec = ~30 million ops

I averaged values for `median`, found total `minimum` and `maximum`, and wrote ranges for `median-absolute-deviation` from the 5 runs.

Before:
```
throughput:
        median= 20095.61 median-absolute-deviation=(144.66-165.23)
        maximum=20890.55 minimum=16236.53
instructions_per_op:
        median= 116948.66 median-absolute-deviation=(16.88-20.84)
        maximum=117274.17 minimum=116579.45
cpu_cycles_per_op:
        median= 57979.33 median-absolute-deviation=(313.56-336.28)
        maximum=71966.78 minimum=55383.71
```

After:
```
throughput:
        median= 20173.58 median-absolute-deviation=(137.35-325.41) # one outlier, next max is 166.18
        maximum=21581.87 minimum=18206.83
instructions_per_op:
        median= 117155.96 median-absolute-deviation=(17.14-21.77)
        maximum=117454.90 minimum=116770.78
cpu_cycles_per_op:
        median= 57666.66 median-absolute-deviation=(302.99-351.00)
        maximum=63078.64 minimum=54617.88
```

Throughput delta: `20173.58 - 20095.61 = +77.97 = +0.388%`
Instructions per op delta: `117155.96 - 116948.66 = +207.30 = +0.177%`
CPU cycles per op delta: `57666.66 - 57979.33 = -312.67 = -0.539%`
CPU cycles per instruction: `0.49222 - 0.49577 = -0.003546 = -0.715%`
